### PR TITLE
[cpp client]support key shared for cpp client

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <ranxiaolong716@gmail.com>

### Motivation

support key_shared for pulsar-cpp-client

### CODE REVIEW

**Comment by wolfstudy** [comment]
@codelipenghui @jiazhai @merlimat @sijie PTAL

**Comment by wolfstudy** [comment]
run integration tests

File: pulsar-client-cpp/include/pulsar/Message.h [hunk 1]
```diff
@@ -110,6 +110,20 @@ class PULSAR_PUBLIC Message {
     const std::string& getPartitionKey() const;
     bool hasPartitionKey() const;
 
+    /**
+     * Get the ordering key of the message
+     *
+     * @return the ordering key of the message
+     */
+    const std::string& getOrderingKey() const;
+    /**
```
Comment (jiazhai) [inline pulsar-client-cpp/include/pulsar/Message.h]: nit: add a new line is better

File: pulsar-client-cpp/tests/ConsumerConfigurationTest.cc [hunk 2]
```diff
@@ -69,6 +69,47 @@ TEST(ConsumerConfigurationTest, testReadCompactPersistentFailover) {
...
+TEST(ConsumerConfigurationTest, testSubscribePersistentKeyShared) {
+    std::string lookupUrl = "pulsar://localhost:6650";
+    std::string topicName = "persist-key-shared-topic";
+    std::string subName = "test-persist-key-shared";
+
+    Result result;
+
+    ConsumerConfiguration config;
+    // now, key-shared not support read compact
+    config.setReadCompacted(false);
```
Comment (jiazhai) [inline pulsar-client-cpp/tests/ConsumerConfigurationTest.cc]: maybe we could remove these 2 lines above

**Comment by wolfstudy** [comment]
run cpp tests
run java8 tests

**Pull request merged**